src/Data/Type/Rec.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Data.Type.Rec
+  ( type Rec,
+    constMap,
+    setRec,
+    getRec,
+    pattern RCon,
+    RecT (RConT, RNil),
+    fieldMap,
+    constMapT,
+    fields,
+    type (.|),
+    HasSel (..),
+
+    -- * Re-exports
+    Ascribe,
+    type (#),
+    Name (Name),
+  )
+where
+
+import Control.Natural (type (~>))
+import Data.Functor.Identity (Identity (Identity, runIdentity))
+import Data.Kind (Constraint, Type)
+import GHC.TypeLits (Symbol)
+
+import Data.Ascript (Ascribe, AscriptName, AscriptType, type (#))
+import Data.Name (Name (Name))
+
+-- -------------------------------------------------------------------------------------------------
+
+type Rec :: [Ascribe Symbol Type] -> Type
+type Rec ctx = RecT Identity ctx
+
+constMap :: Monoid (t a) => (forall s x. Name s -> x -> t a) -> Rec ctx -> t a
+constMap f = constMapT (\name -> f name . Identity)
+{-# INLINE constMap #-}
+
+setRec :: forall ctx s a. s # a .| ctx => Name s -> a -> Rec ctx -> Rec ctx
+setRec name x = setRecT name (Identity x)
+{-# INLINE setRec #-}
+
+getRec :: forall ctx s a. s # a .| ctx => Name s -> Rec ctx -> a
+getRec n r = runIdentity (getRecT n r)
+{-# INLINE getRec #-}
+
+pattern RCon :: () => (s # a ': ctx) ~ ctx' => Name s -> a -> Rec ctx -> Rec ctx'
+pattern RCon n x r <-
+  RConT n (Identity x) r
+  where
+    RCon n x r = RConT n (Identity x) r
+{-# COMPLETE RCon #-}
+
+data RecT f ctx where
+  RNil :: RecT f '[]
+  RConT :: Name s -> f a -> RecT f ctx -> RecT f (s # a ': ctx)
+
+instance Show (RecT f '[]) where
+  show RNil = "RNil"
+
+instance (Show (f a), Show (RecT f ctx)) => Show (RecT f (s # a ': ctx)) where
+  show (RConT name x r) = show name ++ "=" ++ show x ++ " " ++ show r
```

Comment:
Do you think it would be possible (and useful) to add an `INCOHERENT` instance below this with `f ~ Identity` in the instance context?

I imagine it could help avoid poor type inference when you just want to `show` some record.

On a related note, a `Show` instance where `f ~ Identity` (whether in the instance head or context) could be nice, because you could unwrap the `Identity` newtype to avoid the `Identity`  constructor in the `show`n output.

Reply:
I'll add an instance for `Rec`! An overlaps pragma suffices here since `Identity` in the instance head has enough specificity to resolve the choice of show.